torchx/c_src/torchx.cpp

```diff
@@ -21,8 +22,79 @@ inline const std::string* type2string(const torch::ScalarType type)
   return nullptr;
 }
 
+// the class instance to manage the refcount of Tensor
+class TensorP
+{
+public:
+  TensorP(ErlNifEnv *env, const ERL_NIF_TERM arg) : ptr(nullptr)
+  {
+    // setup
+    if (!enif_get_resource(env, arg, TENSOR_TYPE, (void **)&ptr))
+    {
+      // didn't give this parameter
+      return;
+    }
+
+    refcount = (std::atomic<int> *)(ptr + 1);
+
+    if (refcount->load() == 0)
```

Comment:
Are we sure that this can never be negative? Imagine you have three threads: one is holding the tensor on a computation. Now two other threads call deallocate at the same time. Because the tensor is still being used, the tensor is valid, and both deallocate counts will decrease the refcount which can then be negative.

Can this be a thing?

If this is possible, then there is another scenario where two threads are using the tensor, two threads deallocate it, and now the first thread using the tensor will deallocate it, before the other thread using it finishes.

Reply:
this will not protect calling deallocate twice. elixir side will have such issue?

Reply:
but it will not harm if you calling twice. we will still deallocate once on tensor. but both callers will get success message.

Reply:
Yes, you can call deallocate twice. I mean, it is definitely bad code to do so, but it could happen? The issue is that if we call deallocate twice, this condition may never be met because `refcount->load()` will now be negative?

Then, even if we change this to be `refcount->load() < 0`, then there is the issue in that: if you have two threads calling deallocate twice, and then two threads using and blocking a tensor, then when the first thread using and blocking a tensor terminates, it will deallocate before the other thread finishes.

I feel like we have two options:

1. Use a atomic to avoid unsafe writes. If you call deallocate and the refcount is not zero, we raise saying we cannot deallocate. We push the concern to the user, but I think we are fine.

2. Use two atomics: one to track reads and another to track deallocations.

I think we should go with 1 for simplity and we document that, if you try to deallocate while it is being used elsewhere, it will raise. It should not happen in practice anyway. WDYT?

Reply:
Actually, I think my suggestions above are wrong. We need one atomic and one flag to keep it safe:

1. On deallocate, we set the `to_deallocate` flag to false
2. When reading the tensor, we can only bump the refcount if the tensor is not be deallocated
3. When the refcounting is zero and the tensor was set to be deallocated, we effectively deallocate it

The `deallocate` function itself can use the tensor pointer, as all it does it to set a flag to true, and if the refcount is zero at the end, it will deallocate.

Reply:
I don't think this approach will get anything better. The current refcount is a standard implementation. As I mentioned, the only problem is that you will get two successful deallocate call. But deallocate will still happen once.

Reply:
The issue is that if you call deallocate twice, refcount->load() will now return `-1` and this conditional won't match. Setting this clause to `< 0` will also have bugs. I think you need to guarantee deallocation is either idempotent or atomic.

Reply:
This grantee atomic. let me try to explain in graph.

Reply:
![image](https://user-images.githubusercontent.com/766788/159855984-020d6dca-7701-480f-8800-acbb6e08671e.jpeg)

Reply:
Great idea with the drawings. Here is what I am thinking:

```
threadA    threadB    threadC     ref_count
                                      1
 enter                                2
  use
            enter                     3
          deallocate                  2
                       enter          3
                     deallocate       2
            leave                     1
                       leave (!!!)    0
```

In the scenario above we will leave while it is still being used, no?